### PR TITLE
Resize valid_batch_size to avoid OOM.

### PR DESCRIPTION
**Pull request by helson73**
Automatically resizing valid_batch_size to handle OOM.
#492 #476 #301

### CODE REVIEW

**Comment by srush** [comment]
Can you describe the logic here in a comment?  It seems like you are trying to make the size the same for train and valid? 

I still think there might be a memory leak. There is no reason valid should need to use as much memory as train because it doesn't need to hold on to the buffers for backpropagation.

**Comment by helson73** [comment]
@srush 
I think there are 4 buffers for each node. 

- node.Input
- node.Output
- node.gradInput
- node.gradOutput

Of course gradInput and gradOutput are not needed in validation.
But Input and Output buffers are always needed, whether in training or validation.
Can't imagine how a network feedforward signals without using any memory space.
What this patch done is forcing node.Output be same size during training and validation.

Shape of node.Output in training is <max_generator_batches, batch_size, vocab_size>
Shape of node.Output in validation is <max_valid_length, valid_batch_size, vocab_size>

if max_generator_batches * batch_size < max_valid_length * valid_batch_size
then resize valid_batch_size to 
max_generator_batches * batch_size / max_valid_length

**Comment by JianyuZhan** [comment]
@helson73, your reasoning is persuasive, except that I think `volatile` should also means less memory use, quoted pytorch docs: 

> Volatile is recommended for purely inference mode, when you’re sure you won’t be even calling 
> .backward(). It’s more efficient than any other autograd setting - it will use the **absolute minimal 
> amount of memory** to evaluate the model.

Would you also file this problem to upstream? If they claims that `volatile` would save memory but not in fact, then I really suspect there are bugs in code.  I am just feeling this problem should not merely fixed ad-hoc on users side, which would probably hide away some other problems.

Another problem I am worried about is your change now needs to iterate the whole dataset to determine the `max_valid_length`, which would loss the benefit of an `Ordered` iterator, and later 
we will do lazy dataset load on train side to avoid sucking a huge dataset in one go.  `max_valid_length` requires to iterate the whole dataset, and this is a hindrance to that work.

**Comment by helson73** [comment]
@JianyuZhan 
Sorry, my words have some misleading,
What they declared : eval() is much more memory-efficient than train()
But what I want to say is : switch from train() -> eval() doesn't release space of gradients.

I have following code:
```python
import torch
import torch.nn as nn
import torch.optim as optim
from torch.autograd import Variable
import gpustat

gpuid = 2

def show_mem(message):
    stats = gpustat.GPUStatCollection.new_query()
    item = stats.jsonify()["gpus"][gpuid]
    print '%s -- VRAM usage: %d MB' % (message, item["memory.used"])

def create_data():
    show_mem('creating data ... start')
    data = {}
    data['x_train'] = Variable(torch.randn(100,10000).cuda(), volatile=False)
    data['x_test'] = Variable(torch.randn(100,10000).cuda(), volatile=True)
    data['y'] = Variable(torch.randn(100, 10000).cuda(), volatile=False)
    show_mem('creating data ... done')
    return data

def create_net():
    show_mem('creating net ... start')
    net = nn.Linear(10000, 10000)
    net.cuda()
    cri = nn.MSELoss()
    optimer = optim.SGD(net.parameters(), lr=0.1)
    show_mem('creating net ... done')
    return net, cri, optimer

# initializing
data = create_data()
net, cri, optimer = create_net()

# training
show_mem('training ... start')
net.train()
out = net(data['x_train'])
show_mem('training ... after forward')
loss = cri(out, data['y'])
loss.backward()
show_mem('training ... after backward')
optimer.step()
show_mem('training ... done')

# evaluation
show_mem('testing ... start')
net.eval()
out = net(data['x_test'])
show_mem('testing ... done')
```

Got following result:
```
creating data ... start -- VRAM usage: 0 MB
creating data ... done -- VRAM usage: 309 MB
creating net ... start -- VRAM usage: 309 MB
creating net ... done -- VRAM usage: 693 MB
training ... start -- VRAM usage: 693 MB
training ... after forward -- VRAM usage: 731 MB
training ... after backward -- VRAM usage: 1515 MB
training ... done -- VRAM usage: 1515 MB
testing ... start -- VRAM usage: 1515 MB
testing ... done -- VRAM usage: 1515 MB
```

Buffers in node, once it's created, it's size never be changed.

**Comment by helson73** [comment]
@JianyuZhan 
But if I change x_test to:
data['x_test'] = Variable(torch.randn(200,100).cuda(), volatile=True)

Result:
```
creating data ... start -- VRAM usage: 0 MB
creating data ... done -- VRAM usage: 303 MB
creating net ... start -- VRAM usage: 303 MB
creating net ... done -- VRAM usage: 307 MB
training ... start -- VRAM usage: 307 MB
training ... after forward -- VRAM usage: 345 MB
training ... after backward -- VRAM usage: 369 MB
training ... done -- VRAM usage: 369 MB
testing ... start -- VRAM usage: 369 MB
testing ... done -- VRAM usage: 377 MB
```

New space is allocated, because output buffer exceeds.

**Comment by helson73** [comment]
If we use max_valid_length = opt.max_tgt_length, then we don't need to iterate valid dataset, but I am afraid user's validset may be generated from different preprocessing options.
One solution is add max_length field to Dataset object.

**Comment by helson73** [comment]
Also, current implementation of pytorch (keeping gradient buffer) is better choice than dropping them. 
Because pytorch don't know when user want to call eval(), if it happens too frequently, dropping them at eval() and re-allocate new space at train(), could lead to huge lags. 
But it would be nice if we can manually drop them in our case.
